### PR TITLE
Balder/avoid identifiable for valueupdate 2

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli PR
Reduces the use on CloneablePtr and clone()

### CODE REVIEW

**Review by vekterli — APPROVED** [review]
:+1:

File: document/src/vespa/document/update/assignvalueupdate.h [hunk 1]
```diff
@@ -16,16 +16,15 @@
...
+    std::unique_ptr<FieldValue> _value;
 
     ACCEPT_UPDATE_VISITOR;
 public:
-    typedef std::unique_ptr<AssignValueUpdate> UP;
-
     AssignValueUpdate();
-
-    AssignValueUpdate(const FieldValue& value);
+    AssignValueUpdate(std::unique_ptr<FieldValue> value);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/update/assignvalueupdate.h:25]: Fixed

File: document/src/vespa/document/update/removevalueupdate.h [hunk 2]
```diff
@@ -10,22 +10,22 @@
...
-    typedef std::unique_ptr<RemoveValueUpdate> UP;
-
     /**
      * The default constructor requires initial values for all member variables.
      *
      * @param value The identifier of the field value to update.
      */
-    RemoveValueUpdate(const FieldValue& key);
-    ~RemoveValueUpdate();
+    RemoveValueUpdate(std::unique_ptr<FieldValue> key);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/update/removevalueupdate.h:25]: Fixed

File: document/src/vespa/document/update/tensor_add_update.h [hunk 3]
```diff
@@ -14,17 +14,17 @@ class TensorFieldValue;
...
-class TensorAddUpdate : public ValueUpdate, public TensorUpdate {
+class TensorAddUpdate final : public ValueUpdate, public TensorUpdate {
     std::unique_ptr<TensorFieldValue> _tensor;
 
+    friend ValueUpdate;
     TensorAddUpdate();
-    TensorAddUpdate(const TensorAddUpdate &rhs);
     ACCEPT_UPDATE_VISITOR;
 public:
     TensorAddUpdate(std::unique_ptr<TensorFieldValue> &&tensor);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/update/tensor_add_update.h:24]: Yes, fixed

File: document/src/vespa/document/update/tensor_remove_update.cpp [hunk 4]
```diff
@@ -38,53 +38,29 @@ convertToCompatibleType(const TensorDataType &tensorType)
...
-}
-
 TensorRemoveUpdate::TensorRemoveUpdate(std::unique_ptr<TensorFieldValue> tensor)
-    : _tensorType(std::make_unique<TensorDataType>(dynamic_cast<const TensorDataType &>(*tensor->getDataType()))),
+    : ValueUpdate(TensorRemove),
+      TensorUpdate(),
+      _tensorType(std::make_unique<TensorDataType>(dynamic_cast<const TensorDataType &>(*tensor->getDataType()))),
       _tensor(static_cast<TensorFieldValue *>(_tensorType->createFieldValue().release()))
 {
     *_tensor = *tensor;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/update/tensor_remove_update.cpp:55]: Have no idea, will see what complains.

File: document/src/vespa/document/update/valueupdate.cpp [hunk 5]
```diff
@@ -1,29 +1,82 @@
...
+        case Map:
+            return std::unique_ptr<MapValueUpdate>( new MapValueUpdate());
+        case TensorAdd:
+            return std::unique_ptr<TensorAddUpdate>( new TensorAddUpdate());
+        case TensorModify:
+            return std::unique_ptr<TensorModifyUpdate>( new TensorModifyUpdate());
+        case TensorRemove:
+            return std::unique_ptr<TensorRemoveUpdate>( new TensorRemoveUpdate());
+    }
+    throw std::runtime_error(vespalib::make_string("Could not find a class for classId %d(%x)", type, type));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/update/valueupdate.cpp:65]: Fixed

File: document/src/tests/documentselectparsertest.cpp [hunk 6]
```diff
@@ -139,9 +139,9 @@ DocumentUpdate::SP DocumentSelectParserTest::createUpdate(
     const DocumentType* type = _repo->getDocumentType(doctype);
     auto doc = std::make_shared<DocumentUpdate>(*_repo, *type, DocumentId(id));
     doc->addUpdate(FieldUpdate(doc->getType().getField("headerval"))
-                      .addUpdate(std::make_unique<AssignValueUpdate>(IntFieldValue(hint))));
+                      .addUpdate(std::make_unique<AssignValueUpdate>(std::make_unique<IntFieldValue>(hint))));
     doc->addUpdate(FieldUpdate(doc->getType().getField("hstringval"))
-                      .addUpdate(std::make_unique<AssignValueUpdate>(StringFieldValue(hstr))));
+                      .addUpdate(std::make_unique<AssignValueUpdate>(std::make_unique<StringFieldValue>(hstr))));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/tests/documentselectparsertest.cpp:144]: Fixed

File: document/src/vespa/document/serialization/vespadocumentserializer.cpp [hunk 7]
```diff
@@ -377,31 +377,31 @@ VespaDocumentSerializer::write(const FieldUpdate &value)
 void
 VespaDocumentSerializer::write(const RemoveValueUpdate &value)
 {
-    _stream << RemoveValueUpdate::classId;
+    _stream << ValueUpdate::Remove;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/serialization/vespadocumentserializer.cpp]: Proper cast added during serialisation. Deserialisation is already using 32bit int.

**Pull request merged**